libraries/chain/thread_utils.cpp

```diff
@@ -1,35 +1,42 @@
 #include <eosio/chain/thread_utils.hpp>
 #include <fc/log/logger_config.hpp>
+#include <fc/exception/exception.hpp>
 
 namespace eosio { namespace chain {
 
-
-//
-// named_thread_pool
-//
-named_thread_pool::named_thread_pool( std::string name_prefix, size_t num_threads )
-: _thread_pool( num_threads )
+named_thread_pool::named_thread_pool( std::string name_prefix, size_t num_threads, bool delay_start )
+: _name_prefix( std::move(name_prefix) )
+, _num_threads( num_threads )
 , _ioc( num_threads )
 {
+   if( !delay_start ) {
+      start();
+   }
+}
+
+named_thread_pool::~named_thread_pool() {
+   stop();
+}
+
+void named_thread_pool::start() {
+   FC_ASSERT( !_ioc_work, "Thread pool already started" );
    _ioc_work.emplace( boost::asio::make_work_guard( _ioc ) );
-   for( size_t i = 0; i < num_threads; ++i ) {
-      boost::asio::post( _thread_pool, [&ioc = _ioc, name_prefix, i]() {
+   for( size_t i = 0; i < _num_threads; ++i ) {
+      _thread_pool.emplace_back( [&ioc = _ioc, &name_prefix = _name_prefix, i]() {
          std::string tn = name_prefix + "-" + std::to_string( i );
          fc::set_os_thread_name( tn );
          ioc.run();
       } );
    }
 }
 
-named_thread_pool::~named_thread_pool() {
-   stop();
-}
-
 void named_thread_pool::stop() {
    _ioc_work.reset();
    _ioc.stop();
-   _thread_pool.join();
-   _thread_pool.stop();
+   for( auto& t : _thread_pool ) {
+      t.join();
+   }
+   _thread_pool.clear();
```

Comment:
Should those actions be done in the reverse order of `start()`?

Reply:
The order looks correct to me.

Reply:
Though possibly you don't actually need to create & destroy the work guard across start/stops like this.

Reply:
Yeah, maybe, but seems safer to do so.

Reply:
I meant should `_ioc_work.reset();` be done last?

Reply:
It shouldn't really matter, other then it needs to be destroyed before the io_context is destroyed. afaict don't really even need to do the `.reset()` since the class members will be dtored in appropriate order.

Reply:
that's why not making it optional is kinda nice, you can just make the class:
```c++
      boost::asio::io_context        _ioc;
      ioc_work_t      _ioc_work{_ioc);
```
and just not touch it at all.

Reply:
(but I don't think it's something that requires changing)

Reply:
Good point. All the current impl of ioc_work does is call `stop()`.